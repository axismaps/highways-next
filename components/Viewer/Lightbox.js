import React from 'react';
import PropTypes from 'prop-types';
import {
  Box,
  Heading,
  Image,
  Text,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalFooter,
  ModalBody,
  ModalCloseButton,
} from '@chakra-ui/react';

const Lightbox = ({ isOpen, onClose, document }) => (
  <Modal isOpen={isOpen} onClose={onClose} size="2xl">
    <ModalOverlay />
    <ModalContent>
      <ModalHeader />
      <ModalCloseButton />
      <ModalBody>
        <Image
          maxH="75vh"
          maxW="75vw"
          margin="auto"
          src={`${process.env.NEXT_PUBLIC_IMAGE_URL}/${document.ssid}/full.jpg`}
        />
      </ModalBody>
      <ModalFooter justifyContent="flex-start">
        <Box>
          <Heading size="sm">{document.title}</Heading>
          <Text>
            Vivamus tempus ante a interdum pulvinar. Lorem ipsum dolor sit amet, consectetur
            adipiscing elit. Integer.
          </Text>
        </Box>
      </ModalFooter>
    </ModalContent>
  </Modal>
);

Lightbox.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  document: PropTypes.shape().isRequired,
};

export default Lightbox;
