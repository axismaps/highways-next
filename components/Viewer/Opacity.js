import React from 'react';
import PropTypes from 'prop-types';
import { Flex, Text, Slider, SliderTrack, SliderFilledTrack, SliderThumb } from '@chakra-ui/react';

const Opacity = ({ opacityHandler }) => (
  <Flex>
    <Text fontSize={12} fontWeight="bold" mr={5}>
      Opacity:
    </Text>
    <Slider
      aria-label="slider-ex-1"
      defaultValue={1}
      max={1}
      min={0}
      step={0.1}
      onChange={opacityHandler}
      colorScheme="gray"
      mr={2}
    >
      <SliderTrack>
        <SliderFilledTrack />
      </SliderTrack>
      <SliderThumb />
    </Slider>
  </Flex>
);

Opacity.propTypes = {
  opacityHandler: PropTypes.func.isRequired,
};

export default Opacity;
