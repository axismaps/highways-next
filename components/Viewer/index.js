import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import useSWR from 'swr';
import axios from 'axios';
import { Box, Flex, Spacer, Heading, Text, Button, Image } from '@chakra-ui/react';
import { ChevronLeftIcon, ChevronRightIcon, CloseIcon } from '@chakra-ui/icons';

import Opacity from './Opacity';
import Lightbox from './Lightbox';

const fetcher = url => axios.get(url).then(({ data }) => data);
const basemapFetcher = url =>
  axios.get(url).then(({ data: { features } }) => features[0].properties);

const Viewer = ({ year, activeBasemap, opacityHandler, basemapHandler }) => {
  const { data: document, error } = useSWR(
    `${process.env.NEXT_PUBLIC_SEARCH_API}/document/${activeBasemap}`,
    basemapFetcher
  );
  const { data: documents } = useSWR(
    `${process.env.NEXT_PUBLIC_SEARCH_API}/documents?year=${year}`,
    fetcher
  );

  const [lightboxOpen, setLightboxOpen] = useState(false);
  const [type, setType] = useState(null);
  const [currentIndex, setCurrentIndex] = useState(null);

  useEffect(() => {
    if (documents) {
      setType(documents.find(d => d.Documents.find(v => v.ssid === activeBasemap)));
    }
  }, [documents, activeBasemap]);

  useEffect(() => {
    if (type) {
      setCurrentIndex(type.Documents.findIndex(t => t.ssid === activeBasemap));
    }
  }, [type, activeBasemap]);

  const changeView = step => {
    let nextIndex = currentIndex + step;
    if (nextIndex >= type.Documents.length) nextIndex = 0;
    else if (nextIndex < 0) nextIndex = type.Documents.length - 1;
    basemapHandler(type.Documents[nextIndex].ssid);
  };

  if (!document || !documents || error) return 'LOADING';

  return (
    <Box
      pos="absolute"
      right={[0, '10px']}
      top={['auto', '120px']}
      bottom={[0, 'auto']}
      w={['100%', '330px']}
      p="20px"
      gridTemplateRows="1fr 1fr 1fr 40px"
      rowGap="20px"
      backgroundColor="white"
      zIndex={8}
    >
      <Flex mb={2}>
        <Heading size="sm" fontSize={[14, 'auto']}>
          {document.title}
        </Heading>
        <Spacer />
        <CloseIcon onClick={() => basemapHandler(null)} fontSize={12} mt={1} cursor="pointer" />
      </Flex>
      <Box display={['flex', 'block']} mb={[2, 0]}>
        <Box pos="relative" mb={2} maxWidth={['25%', '100%']}>
          <Image
            src={`${process.env.NEXT_PUBLIC_IMAGE_URL}/${document.ssid}/medium.jpg`}
            onClick={() => setLightboxOpen(true)}
          />
        </Box>
        {document.creator && (
          <Text mb={[0, 2]} ml={[2, 0]} fontSize={['0.75em', 'auto']}>
            {document.creator}
          </Text>
        )}
      </Box>
      {type.title.match(/view/gi) ? (
        <Flex>
          <Button leftIcon={<ChevronLeftIcon />} w="40%" onClick={() => changeView(-1)}>
            Prev
          </Button>
          <Spacer />
          <Button rightIcon={<ChevronRightIcon />} w="40%" onClick={() => changeView(1)}>
            Next
          </Button>
        </Flex>
      ) : (
        <Opacity opacityHandler={opacityHandler} />
      )}
      <Lightbox isOpen={lightboxOpen} onClose={() => setLightboxOpen(false)} document={document} />
    </Box>
  );
};

Viewer.propTypes = {
  year: PropTypes.number.isRequired,
  activeBasemap: PropTypes.string.isRequired,
  opacityHandler: PropTypes.func.isRequired,
  basemapHandler: PropTypes.func.isRequired,
};

export default Viewer;
