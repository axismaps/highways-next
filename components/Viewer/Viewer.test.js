import React from 'react';
import axios from 'axios';
import renderer from 'react-test-renderer';
import MockAdapter from 'axios-mock-adapter';
import { render, screen } from '../../test/test-utils';

import Viewer from './index';
import document from '../../__mocks__/document.data';
import documents from '../../__mocks__/documents.data';

describe('Viewer', () => {
  const props = {
    activeBasemap: 'SSID25009564',
    basemapHandler: () => {},
    opacityHandler: () => {},
    year: 1950,
  };
  const mock = new MockAdapter(axios);
  mock.onGet(`${process.env.NEXT_PUBLIC_SEARCH_API}/documents?year=1950`).reply(200, documents);
  mock.onGet(`${process.env.NEXT_PUBLIC_SEARCH_API}/document/SSID25009564`).reply(200, document);

  it('Renders the viewer image', async () => {
    render(<Viewer {...props} />);
    await screen.findAllByText(text => text.match(/^Sanborn/));
    expect(screen.getAllByText('Sanborn Map Houston 1950 Vol 15 Key 02')[0]).toBeInTheDocument();
  });

  it('matches snapshot', async () => {
    render(<Viewer {...props} />);
    await screen.findAllByText(text => text.match(/^Sanborn/));

    const tree = renderer.create(<Viewer {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
