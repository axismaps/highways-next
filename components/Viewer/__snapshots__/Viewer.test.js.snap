// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewer matches snapshot 1`] = `
<div
  className="css-revwgf"
>
  <div
    className="css-1jqq95v"
  >
    <h2
      className="chakra-heading css-1kuy7z7"
    >
      Sanborn Map Houston 1950 Vol 15 Key 02
    </h2>
    <div
      className="css-17xejub"
    />
    <svg
      className="chakra-icon css-1gyjhah"
      focusable={false}
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
        fill="currentColor"
      />
    </svg>
  </div>
  <div
    className="css-1jqq95v"
  >
    <div
      className="css-1i2pf6n"
    >
      <img
        className="chakra-image__placeholder css-0"
        onClick={[Function]}
      />
    </div>
    <p
      className="chakra-text css-1bmsndy"
    >
      Sanborn Perris Map Company
    </p>
  </div>
  <div
    className="css-k008qs"
  >
    <p
      className="chakra-text css-jrrolj"
    >
      Opacity:
    </p>
    <div
      className="chakra-slider css-1rmi0bl"
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0,0,0,0)",
          "outline": 0,
          "paddingBottom": 0,
          "paddingTop": 0,
          "position": "relative",
          "touchAction": "none",
          "userSelect": "none",
        }
      }
      tabIndex={-1}
    >
      <div
        className="chakra-slider__track css-i6bazn"
        id="slider-track-7"
        style={
          Object {
            "position": "absolute",
            "top": "50%",
            "transform": "translateY(-50%)",
            "width": "100%",
          }
        }
      >
        <div
          className="chakra-slider__filled-track css-p8tu8l"
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": "50%",
              "transform": "translateY(-50%)",
              "width": "100%",
            }
          }
        />
      </div>
      <div
        aria-label="slider-ex-1"
        aria-orientation="horizontal"
        aria-valuemax={1}
        aria-valuemin={0}
        aria-valuenow={1}
        className="chakra-slider__thumb css-lmsxnt"
        id="slider-thumb-7"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="slider"
        style={
          Object {
            "left": "calc(100% - 0px)",
            "position": "absolute",
            "touchAction": "none",
            "userSelect": "none",
          }
        }
        tabIndex={0}
      />
      <input
        type="hidden"
        value={1}
      />
    </div>
  </div>
</div>
`;
